### PR TITLE
feat(Genshin Impact Map): add presence

### PR DESCRIPTION
**Pull request by Faelayis**
**Describe the changes in this pull request:**

<details>
<summary> Proof showing the creation/modification is working as expected </summary>
<br>

image
image
image
image
image
image
image
image

</details>

### CODE REVIEW

File: websites/G/Genshin Impact Map/dist/metadata.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	},
+	"url": [
+		"webstatic-sea.mihoyo.com",
+		"webstatic-sea.hoyolab.com",
+		"genshin-impact-map.appsample.com",
+		"mapgenie.io"
+	],
+	"version": "1.0.0",
+	"logo": "https://i.imgur.com/6XrkcZs.png",
+	"thumbnail": "https://i.imgur.com/6DmC0EO.png",
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/dist/metadata.json:19]: please try to find a wider image, preferrable 16:9

File: websites/G/Genshin Impact Map/presence.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
+const presence = new Presence({
+		clientId: "973729201104511007"
+	}),
+	browsingTimestamp = Math.floor(Date.now() / 1000),
+	maps = [
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:5]: this array should be typed

File: websites/G/Genshin Impact Map/presence.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+							document
+								.querySelector("#map-selector-btn > strong > p")
+								.textContent.toLowerCase()
+						)
+				) ??
+				maps.find(
+					i =>
+						i.map.toLowerCase().includes(path[3].split("?map=")[1]) ||
+						i.hylid === 2
+				);
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:105]: Lint will probably fail with this so make sure to run it afterwards
```suggestion
				maps.find(i =>
					i.map
						.toLowerCase()
						.includes(
							document
								.querySelector("#map-selector-btn > strong > p")
								.textContent.toLowerCase()
						) ||
						i.map.toLowerCase().includes(path[3].split("?map=")[1]) ||
						i.hylid === 2
				);
```

File: websites/G/Genshin Impact Map/presence.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+let getpos: number,
+	current: {
+		city: boolean;
+		hylid: number;
+		map: string;
+		largeImageKey: string;
+		pvlargeImageKey: string | null;
+		smallImageKey: string;
+	},
+	currentcity: {
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:70]:
```suggestion
	currentCity: {
```

File: websites/G/Genshin Impact Map/dist/metadata.json [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	"color": "#b6c48c",
+	"category": "games",
+	"tags": [
+		"genshin-impact",
+		"map"
+	],
+	"settings": [
+		{
+			"id": "showepreview",
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/dist/metadata.json:28]:
```suggestion
			"id": "showPreview",
```

File: websites/G/Genshin Impact Map/presence.ts [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+		position?: number[];
+		map: string;
+		largeImageKey: string;
+		smallImageKey: string;
+	};
+
+presence.on("UpdateData", async () => {
+	const [showepreview, timestamps] = await Promise.all([
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:78]:
```suggestion
	const [showPreview, timestamps] = await Promise.all([
```

File: websites/G/Genshin Impact Map/presence.ts [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+			presence.getSetting<boolean>("showepreview"),
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:79]:
```suggestion
			presence.getSetting<boolean>("showPreview"),
```

**Review by ImRodry — CHANGES_REQUESTED** [review]
Definitely an interesting approach to a presence, good job! Just some small nits

File: websites/G/Genshin Impact Map/presence.ts (hunk 2 above)
Comment (Faelayis) [reply websites/G/Genshin Impact Map/presence.ts:5]: maps top browsingTimestamp ?

Comment (ImRodry) [reply websites/G/Genshin Impact Map/presence.ts:5]: what

Comment (Faelayis) [reply websites/G/Genshin Impact Map/presence.ts:5]: where array typed ?

Comment (ImRodry) [reply websites/G/Genshin Impact Map/presence.ts:5]: I’m saying the array should be typed with an interface along with the variables below

File: websites/G/Genshin Impact Map/presence.ts [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+			smallImageKey: "emblem_mondstadt"
+		},
+		{
+			position: [4000],
+			map: "Liyue",
+			largeImageKey: "preview_liyue",
+			smallImageKey: "emblem_liyue"
+		},
+		{
+			position: [9000],
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:53]: why are these arrays with one element only

File: websites/G/Genshin Impact Map/presence.ts [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+interface maps {
+	city: boolean;
+	hylid: number;
+	map: string;
+	largeImageKey: string;
+	pvlargeImageKey: string | null;
+	smallImageKey: string;
+}
+
+interface city {
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:70]:
```suggestion
interface Maps {
	city: boolean;
	hylid: number;
	map: string;
	largeImageKey: string;
	pvlargeImageKey: string | null;
	smallImageKey: string;
}

interface City {
```

**Review by ImRodry — CHANGES_REQUESTED** [review]
Please address the remaining review

File: websites/G/Genshin Impact Map/presence.ts [hunk 10]
```diff
@@ -0,0 +1,166 @@
+const presence = new Presence({
+		clientId: "973729201104511007"
+	}),
+	browsingTimestamp = Math.floor(Date.now() / 1000);
+
+interface Maps {
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:6]:
```suggestion
interface Map {
```

File: websites/G/Genshin Impact Map/presence.ts [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+		{
+			city: false,
+			hylid: 0,
+			map: "Unknown",
+			largeImageKey: "unknown_map",
+			pvlargeImageKey: null,
+			smallImageKey: "emblem_unknown"
+		}
+	],
+	City = [
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:56]:
```suggestion
	city: City[] = [
```

File: websites/G/Genshin Impact Map/presence.ts [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+			largeImageKey: "preview_liyue",
+			smallImageKey: "emblem_liyue"
+		},
+		{
+			position: 9000,
+			map: "Inazuma",
+			largeImageKey: "preview_tenshukaku",
+			smallImageKey: "emblem_inazuma"
+		}
+		// { position: [], map: "Sumeru", largeImageKey: "", smallImageKey: "" }
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:75]: why is this here?

File: websites/G/Genshin Impact Map/presence.ts [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+		? currentCity
+			? currentCity.smallImageKey
+			: current.smallImageKey
+		: current.smallImageKey;
+	presenceData.smallImageText = document.location.host.replace(".com", "");
+	if (!timestamps) {
+		delete presenceData.startTimestamp;
+		delete presenceData.endTimestamp;
+	}
+	console.log(currentCity);
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:163]: delete this

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: websites/G/Genshin Impact Map/presence.ts (hunk 10 above)
Comment (Faelayis) [reply websites/G/Genshin Impact Map/presence.ts:6]: cannot ,don't change it any more u make it inactive

Comment (ImRodry) [reply websites/G/Genshin Impact Map/presence.ts:6]: you have to change this in your IDE along with all the references to it, you're not meant to merge this suggestion simply

**Comment by ImRodry** [comment]
Please undo the changes to yarn.lock. Also interfaces should be PascalCased, which is not what you did.

File: websites/G/Genshin Impact Map/presence.ts [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+		{
+			city: false,
+			hylid: 0,
+			map: "Unknown",
+			largeImageKey: "unknown_map",
+			pvlargeImageKey: null,
+			smallImageKey: "emblem_unknown"
+		}
+	],
+	City: City[] = [
```
Comment (Slowlife01) [inline websites/G/Genshin Impact Map/presence.ts:56]:
```suggestion
const map: Map[] = [
		{
			city: true,
			hylid: 2,
			map: "Teyvat",
			largeImageKey: "teyvat_map",
			pvlargeImageKey: "teyvat_map",
			smallImageKey: "emblem_unknown"
		},
		{
			city: false,
			hylid: 7,
			map: "Enkanomiya",
			largeImageKey: "enkanomiya_map",
			pvlargeImageKey: "preview_enkanomiya",
			smallImageKey: "emblem_enkanomiya"
		},
		{
			city: false,
			hylid: 9,
			map: "The Chasm: Underground Mines",
			largeImageKey: "the_chasm_underground_mines_map",
			pvlargeImageKey: "preview_the_chasm_underground_mines",
			smallImageKey: "emblem_thechasm"
		},
		{
			city: false,
			hylid: 0,
			map: "Unknown",
			largeImageKey: "unknown_map",
			pvlargeImageKey: null,
			smallImageKey: "emblem_unknown"
		}
	],
	city: City[] = [
```

**Review by EncryptedDev — APPROVED** [review]
LGTM

**Review by Slowlife01 — APPROVED** [review]

**Review by Slowlife01 — APPROVED** [review]

File: websites/G/Genshin Impact Map/presence.ts [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+			presence.getSetting<boolean>("showPreview"),
+			presence.getSetting<boolean>("timestamps")
+		]),
+		presenceData: PresenceData = {
+			details: "Genshin Impact Map",
+			largeImageKey: "main",
+			smallImageKey: "search",
+			startTimestamp: browsingTimestamp
+		},
+		path = document.location.toString().split("/");
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:90]:
```suggestion
		path = document.location.href.split("/");
```

File: websites/G/Genshin Impact Map/presence.ts [hunk 16]
```diff
@@ -0,0 +1,163 @@
+const presence = new Presence({
+		clientId: "973729201104511007"
+	}),
+	browsingTimestamp = Math.floor(Date.now() / 1000);
+
+interface Map {
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:6]: Map is a node class, you shouldn't be shadowing global definitions so you should probably name this something different

**Review by ImRodry — CHANGES_REQUESTED** [review]
Your `path` variable is only used once and it's equal to document.location.href which is used a lot more often, I would recommend making a variable that fits all of these use cases to be more efficient

**Comment by ImRodry** [comment]
@Faelayis awaiting reply

**Comment by Faelayis** [comment]
> @Faelayis awaiting reply

what ?

File: websites/G/Genshin Impact Map/presence.ts (hunk 10 above)
Comment (EncryptedDev) [reply websites/G/Genshin Impact Map/presence.ts:6]: This has not been addressed

File: websites/G/Genshin Impact Map/presence.ts (hunk 16 above)
Comment (EncryptedDev) [reply websites/G/Genshin Impact Map/presence.ts]: This should be addressed

**Comment by ImRodry** [comment]
> > @Faelayis awaiting reply
> 
> what ? ,halooo

I asked you a question, I'm waiting for an answer otherwise this isn't getting merged

**Comment by Faelayis** [comment]
> > > @Faelayis awaiting reply
> > 
> > 
> > what ? ,halooo
> 
> I asked you a question, I'm waiting for an answer otherwise this isn't getting merged

what question ?

**Comment by ImRodry** [comment]
> Your `path` variable is only used once and it's equal to document.location.href which is used a lot more often, I would recommend making a variable that fits all of these use cases to be more efficient

**Comment by ImRodry** [comment]
Rebased to apply the latest changes in the prettier configuration. If you need help with your local branch, please ping me on the #presence-dev channel on Discord.

File: websites/G/Genshin Impact Map/presence.ts [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+			presence.getSetting<boolean>("showPreview"),
+			presence.getSetting<boolean>("timestamps"),
+		]),
+		presenceData: PresenceData = {
+			details: "Genshin Impact Map",
+			largeImageKey: "main",
+			smallImageKey: "search",
+			startTimestamp: browsingTimestamp,
+		},
+		path = document.location.href;
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:90]: If you want the path call it pathname and use the pathname property. If you want the href call it href and keep it like this. Don’t mix them up please

File: websites/G/Genshin Impact Map/presence.ts [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+			break;
+		default: // Official Site
+			if (!path.includes("&center=") && !path.includes("&zoom=")) return;
+			getpos = parseInt(
+				path.split("&center=")[1].split("&zoom=")[0].split(",")[0]
+			);
+			current = map.find(
+				i => i.hylid === parseInt(path.split("/map/")[1].split("?")[0]) || 0
+			);
+			if (current.city) currentCity = city.find(i => i.position > getpos);
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:122]: Can’t current be undefined here? Should probably add optional chaining

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: websites/G/Genshin Impact Map/presence.ts [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+			presence.getSetting<boolean>("showPreview"),
+			presence.getSetting<boolean>("timestamps"),
+		]),
+		presenceData: PresenceData = {
+			details: "Genshin Impact Map",
+			largeImageKey: "main",
+			smallImageKey: "search",
+			startTimestamp: browsingTimestamp,
+		},
+		{ host, hostname, href } = document.location;
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:90]: Seems like you didn’t get the hint so I’ll be more specific. There’s no reason for you to use the href here. You should be using search and pathname

File: websites/G/Genshin Impact Map/presence.ts [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+			current = map.find(
+				i =>
+					i.map
+						.toLowerCase()
+						.includes(
+							document
+								.querySelector("#map-selector-btn > strong > p")
+								.textContent.toLowerCase()
+						) ||
+					i.map.toLowerCase().includes(pathname.split("?map=")[1] || "Teyvat")
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:102]: like I said, you should use `search` here. Please test your changes before committing

File: websites/G/Genshin Impact Map/presence.ts [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+			else {
+				current = map.find(i =>
+					i.map
+						.toLowerCase()
+						.includes(pathname.split("/maps/")[1].toLowerCase())
+				);
+			}
+			break;
+		default: // Official Site
+			if (!pathname.includes("&center=") && !pathname.includes("&zoom="))
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:117]: same here

File: websites/G/Genshin Impact Map/presence.ts [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+				return;
+			getpos = parseInt(
+				pathname.split("&center=")[1].split("&zoom=")[0].split(",")[0]
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:120]: and here

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: websites/G/Genshin Impact Map/presence.ts (hunk 20 above)
Comment (ImRodry) [reply websites/G/Genshin Impact Map/presence.ts:102]: and why does it not?

Comment (Faelayis) [reply websites/G/Genshin Impact Map/presence.ts:102]: i don't know it's not error but no res

File: websites/G/Genshin Impact Map/presence.ts [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+			else {
+				current = map.find(i =>
+					i.map
+						.toLowerCase()
+						.includes(pathname?.split("/maps/")[1].toLowerCase() || "teyvat")
+				);
+			}
+			break;
+		default: // Official Site
+			if (!hash.includes("&center=") && !hash.includes("&zoom=")) return;
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:111]: hash? Did you test this? Please show me a screenshot of a URL where this works

File: websites/G/Genshin Impact Map/presence.ts [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+			getpos = parseInt(
+				hash.split("&center=")[1].split("&zoom=")[0].split(",")[0]
```
Comment (ImRodry) [inline websites/G/Genshin Impact Map/presence.ts:113]: You very likely want to use search here. You should use URLSearchParams

File: websites/G/Genshin Impact Map/presence.ts (hunk 23 above)
Comment (Faelayis) [reply websites/G/Genshin Impact Map/presence.ts:111]: yes i test ,it works

Comment (Faelayis) [reply websites/G/Genshin Impact Map/presence.ts:111]: image

**Review by ImRodry — APPROVED** [review]
If it works it works
